web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java

```diff
@@ -0,0 +1,184 @@
+package dev.langchain4j.web.search.searchapi;
+
+import static dev.langchain4j.internal.Utils.getOrDefault;
+import static dev.langchain4j.internal.Utils.isNotNullOrBlank;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotNull;
+import static java.time.Duration.ofSeconds;
+import static java.util.stream.Collectors.toList;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import dev.langchain4j.web.search.WebSearchEngine;
+import dev.langchain4j.web.search.WebSearchInformationResult;
+import dev.langchain4j.web.search.WebSearchOrganicResult;
+import dev.langchain4j.web.search.WebSearchRequest;
+import dev.langchain4j.web.search.WebSearchResults;
+import dev.langchain4j.web.search.searchapi.result.OrganicResult;
+import lombok.Builder;
+
+/**
+ * Represents SearchApi Search API as a {@code WebSearchEngine}. See more
+ * details <a href="https://www.searchapi.io/docs/google">here</a>.
+ */
+public class SearchApiWebSearchEngine implements WebSearchEngine {
+	public static final String DEFAULT_ENV_VAR = "SEARCHAPI_API_KEY";
+	public static final String DEFAULT_BASE_URL = "https://www.searchapi.io";
+	public static final String DEFAULT_ENGINE = "google";
+	private static final Integer DEFAULT_MAX_RESULTS = 5;
+
+	private final String apiKey;
+	private final String engine;
+	private final SearchApiClient searchapiClient;
+	private final Boolean logRequests;
+
+	@Builder
+	public SearchApiWebSearchEngine(final String apiKey, final String engine, final Duration timeout,
+			final Boolean logRequests) {
+
+		this.apiKey = ensureNotBlank(apiKey, "apiKey");
+		this.engine = ensureNotBlank(getOrDefault(engine, DEFAULT_ENGINE), "engine");
+		this.logRequests = logRequests;
+
+		this.searchapiClient = SearchApiClient.builder().baseUrl(DEFAULT_BASE_URL).apiKey(this.apiKey)
+				.engine(this.engine).timeout(getOrDefault(timeout, ofSeconds(10))).logRequests(this.logRequests)
+				.build();
+	}
+
+	@Override
+	public WebSearchResults search(final WebSearchRequest webSearchRequest) {
+		ensureNotNull(webSearchRequest, "webSearchRequest");
+
+		final SearchApiRequest.SearchApiRequestBuilder requestBuilder = SearchApiRequest.builder();
+
+		requestBuilder.q(webSearchRequest.searchTerms()).safe(webSearchRequest.safeSearch())
+				.num(getOrDefault(webSearchRequest.maxResults(), DEFAULT_MAX_RESULTS)) // maxResults should default to 5
+				.page(webSearchRequest.startPage());
+
+		/**
+		 * {@link WebSearchRequest#language} is in a composite string "en-us" but
+		 * SearchApi splits them into 2 individual strings. See
```

Comment:
please extract this logic into separate static method and cover it with unit tests

Reply:
I decided to extract the logic into a separate static class (instead of a static method).
* [`Locale`](https://github.com/ayewo/langchain4j/blob/d44c28591ad10d65ddef6330b93066eeece3c6f2/web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java#L170-L193)

As requested, I added unit tests: 
* [`LocaleTest`](https://github.com/ayewo/langchain4j/blob/952e0169558c6c2903450f167eec6fa61c478572/web-search-engines/langchain4j-web-search-engine-searchapi/src/test/java/dev/langchain4j/web/search/searchapi/LocaleTest.java)